### PR TITLE
fix(query): prevent creating QueryObserver multiple times

### PR DESCRIPTION
**Pull request by Aslemammad**
Resolves: #957

This issue tries to track the query options through `queryObserver.setOptions`, so @tanstack/query-core features like `keepPreviousData` work properly!

### CODE REVIEW

File: src/query/atomWithQuery.ts [hunk 1]
```diff
@@ -156,7 +170,6 @@ export function atomWithQuery<
             }
           }, 1000)
           resolve(result)
-          resolve = null
```
Comment (dai-shi) [inline src/query/atomWithQuery.ts:159]: This seems bad. (Interesting all tests pass... 🤔

File: src/query/atomWithQuery.ts [hunk 2]
```diff
@@ -123,8 +130,15 @@ export function atomWithQuery<
         typeof createQuery === 'function' ? createQuery(get) : createQuery
 
       const defaultedOptions = queryClient.defaultQueryOptions(options)
-      const observer = new QueryObserver(queryClient, defaultedOptions)
-      const initialResult = observer.getCurrentResult()
+
+      if (observer) {
```
Comment (dai-shi) [inline src/query/atomWithQuery.ts:134]: This doesn't work if `queryClient` is changed, right?

File: src/query/atomWithQuery.ts [hunk 3]
```diff
@@ -95,6 +94,14 @@ export function atomWithQuery<
     isMounted: boolean
     unsubscribe: (() => void) | null
   }
+
+  let observer: QueryObserver<
```
Comment (dai-shi) [inline src/query/atomWithQuery.ts:98]: this is going to be a module scope variable, which breaks the use case with multiple `<Provider>`s.

File: src/query/atomWithQuery.ts (hunk 2 above)
Comment (Aslemammad) [reply src/query/atomWithQuery.ts:134]: I cannot say yes or no! but in useQuery, it tracks the queryClient!

File: src/query/atomWithQuery.ts (hunk 3 above)
Comment (Aslemammad) [reply src/query/atomWithQuery.ts:98]: Isn't refAtom the best solution for that now?

File: src/query/atomWithQuery.ts (hunk 1 above)
Comment (Aslemammad) [reply src/query/atomWithQuery.ts:159]: Removing it will make the tests fail!

Comment (Aslemammad) [reply src/query/atomWithQuery.ts:159]: in versionedWrite

File: src/query/atomWithQuery.ts (hunk 3 above)
Comment (dai-shi) [reply src/query/atomWithQuery.ts:98]: It's better, but not best. I still believe there's a way without refAtom.

File: src/query/atomWithQuery.ts (hunk 1 above)
Comment (dai-shi) [reply src/query/atomWithQuery.ts:159]: hmm, there might be something wrong in other places..

File: src/query/atomWithQuery.ts [hunk 4]
```diff
@@ -90,29 +90,39 @@ export function atomWithQuery<
...
+    TError,
+    TData,
+    TQueryData,
+    TQueryKey
+  >
   type Result = QueryObserverResult<TData, TError>
-  type State = {
-    isMounted: boolean
-    unsubscribe: (() => void) | null
-  }
```
Comment (dai-shi) [inline src/query/atomWithQuery.ts:97]: Good that I was able to remove this. Didn't like this.

File: src/query/atomWithQuery.ts (hunk 1 above)
Comment (dai-shi) [reply src/query/atomWithQuery.ts:159]: We don't have a test to catch if we remove `resolve = null`.
It's a test to update result from queryClient, I suppose.

**Review by dai-shi — APPROVED** [review]
Should be good to go.

**Review by Aslemammad — COMMENTED** [review]
LGTM!

**Pull request merged**